From 687907130adf1f3e1a076c24522f1b7ab2c0466a Mon Sep 17 00:00:00 2001
From: kenzyun_chen <kenzyun_chen@htc.com>
Date: Fri, 17 Oct 2014 15:26:13 +0800
Subject: [PATCH 183/296] Handle IllegalStateException from Surface.lockCanvas

Symptom: SystemServer crash because of this exception on startingWindow
Root Cause: IllegalStateException from Surface.lockCanvas
Solution: handle IllegalStateException from Surface.lockCanvas
---
 core/java/android/view/ViewRootImpl.java | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/core/java/android/view/ViewRootImpl.java b/core/java/android/view/ViewRootImpl.java
index 0023c4f2053..09acbf4402d 100644
--- a/core/java/android/view/ViewRootImpl.java
+++ b/core/java/android/view/ViewRootImpl.java
@@ -3074,6 +3074,12 @@ public final class ViewRootImpl implements ViewParent,
             // kill stuff (or ourself) for no reason.
             mLayoutRequested = true;    // ask wm for a new surface next time.
             return false;
+        } catch (IllegalStateException e) {
+            // After queueBuffer has been abandoned, Surface.unlockCanvasAndPost throws IllegalArgumentException.
+            // However, mLockedObject is not clear in Surface.
+            // This will lead to IllegalStateException while calling Surface.lockCanvas.
+            Log.e(TAG, "Could not lock surface after unlockCanvasAndPost failed", e);
+            return false;
         }
 
         try {
-- 
2.11.0

